usrsctplib/user_environment.c

```diff
@@ -85,6 +85,31 @@ read_random(void *buf, size_t size)
 	return;
 }
 
+void
+finish_random(void)
+{
+	return;
+}
+#elif defined(SCTP_USE_OPENSSL_RAND)
+#include <openssl/rand.h>
+
+void
+init_random(void)
+{
+	return;
+}
+
+void
+read_random(void *buf, size_t size)
+{
+	while (!RAND_bytes((uint8_t *)buf, size)) {
+#ifdef INVARIANTS
+		panic("RAND_bytes failed.");
+#else
+	}
+	return;
```

Comment:
If this function doesn't abort() in case it detects a random number generator issue (when INVARIANTS is not defined), I think it should provide the caller a way to detect the problem (e.g. via return value?).

As of now it silently fails if RAND_bytes returns -1 (https://www.openssl.org/docs/man1.1.1/man3/RAND_bytes.html) which is a failure mode.

Reply:
As I said: the model is that the library does not terminate the program...

Reply:
An infinite loop doesn't seem much preferable to terminating, but if that's what you're doing everywhere else, makes sense to be consistent.

Reply:
I think the infinite loop should be avoided. And it is not used everywhere (e.g. see the Fuchsia version https://github.com/sctplab/usrsctp/blob/c65538fbe14e0141db5053bc24d7cf7d8e2ed8fc/usrsctplib/user_environment.c#L312 or the FreeBSD one https://github.com/sctplab/usrsctp/blob/c65538fbe14e0141db5053bc24d7cf7d8e2ed8fc/usrsctplib/user_environment.c#L103).

If you `#ifndef` on `OPENSSL_IS_BORINGSSL ` (see the other comment), maybe you can remove the loop and call directly like it happens for other flavours?

Reply:
Makes sense to me...

Reply:
Done.

Reply:
In terms of infinite loops, note the other `read_random` functions all have infinite loop issues. E.g. if opening /dev/urandom fails, `read_random` will just loop forever trying to read from -1. In general, with the exception of EINTR, I wouldn't expect retrying the operation to make progress. That means the library can either abort or return an error up to the caller.

If this project doesn't like to abort, the other option would be making `read_random` return an error. Then everything up the call stack would need to handle that error.

Reply:
> In terms of infinite loops, note the other `read_random` functions all have infinite loop issues. E.g. if opening /dev/urandom fails, `read_random` will just loop forever trying to read from -1. In general, with the exception of EINTR, I wouldn't expect retrying the operation to make progress. That means the library can either abort or return an error up to the caller.
> 
> If this project doesn't like to abort, the other option would be making `read_random` return an error. Then everything up the call stack would need to handle that error.

I'm happy to accept patches, but all changes should be `#ifdef`ed in a way that only the userland stack is affected, since in the kernel the function can not fail. You might need to drop packets if the problem happens in the context of handling received packets...